### PR TITLE
8334371: [AIX] Beginning with AIX 7.3 TL1 mmap() supports 64K memory pages

### PR DESCRIPTION
**Pull request by JoKern65**
Beginning with AIX 7.3 TL1 mmap() supports 64K memory pages. As an enhancement, during the initialization of the VM the availability of this new feature is examined. If the 64K pages are supported the VM will use mmap() with 64K pages instead of shmget()/shmat() with 64K pages due to the bad 256M alignment of shmget()/shmat().

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334371](https://bugs.openjdk.org/browse/JDK-8334371): [AIX] Beginning with AIX 7.3 TL1 mmap() supports 64K memory pages (**Enhancement** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) 🔄 Re-review required (review applies to [cd15f133](https://git.openjdk.org/jdk/pull/19771/files/cd15f13396d6a927149a030d5d36f82f7d9b7e37))
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19771/head:pull/19771` \
`$ git checkout pull/19771`

Update a local copy of the PR: \
`$ git checkout pull/19771` \
`$ git pull https://git.openjdk.org/jdk.git pull/19771/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19771`

View PR using the GUI difftool: \
`$ git pr show -t 19771`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19771.diff">https://git.openjdk.org/jdk/pull/19771.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19771#issuecomment-2176165072)

### CODE REVIEW

File: src/hotspot/os/aix/os_aix.cpp [hunk 1]
```diff
@@ -448,6 +453,14 @@ static void query_multipage_support() {
       }
     }
 
+    // Can we use mmap with 64K pages? (Should be available with AIX7.3 TL1)
+    {
+      void* p = mmap(NULL, 1000000, PROT_READ | PROT_WRITE, MAP_ANON_64K | MAP_ANONYMOUS | MAP_SHARED, -1, 0);
```
Comment (tstuefe) [inline src/hotspot/os/aix/os_aix.cpp:458]: Weird size. Why 1mio decimal? Why not 64K?

File: src/hotspot/os/aix/os_aix.cpp [hunk 2]
```diff
@@ -448,6 +453,14 @@ static void query_multipage_support() {
...
+      guarantee0(p != (void*) -1); // Should always work.
```
Comment (tstuefe) [inline src/hotspot/os/aix/os_aix.cpp:459]: I probably introduced those guarantee's myself many years ago, but we actually used guarantee very rarely, only for things that absolutely have to work, or the world explodes. 

Here, there are a number of reasons why an mmap could fail, and you probably don't want to crash the VM with an guarantee at a customer site. I'd use assert.

File: test/hotspot/gtest/runtime/test_os.cpp [hunk 3]
```diff
@@ -38,6 +38,13 @@
 #ifdef _WIN32
 #include "os_windows.hpp"
 #endif
+#ifdef AIX
+# include <sys/mman.h>
+// sys/mman.h defines MAP_ANON_64K beginning with AIX7.3 TL1
+# ifndef MAP_ANON_64K
+# define MAP_ANON_64K  0x400
+# endif
+#endif
```
Comment (tstuefe) [inline test/hotspot/gtest/runtime/test_os.cpp:47]: Could you move the AIX-specific files out to something like test_os_aix.cpp, please? (same directory, just use ifdef AIX in the whole file). Its nicer, and precedence is there with the Cgroup tests of linux.

**Review by tstuefe — COMMENTED** [review]
At long last. Cool.

Once 7.3 is the minimum requirement, you can shake lose a lot of code here, all that old SystemV shm coding.

Small remarks, otherwise okay

File: src/hotspot/os/aix/os_aix.cpp (hunk 2 above)
Comment (JoKern65) [reply src/hotspot/os/aix/os_aix.cpp:459]: OK, I will use 64*K

File: src/hotspot/os/aix/os_aix.cpp (hunk 1 above)
Comment (JoKern65) [reply src/hotspot/os/aix/os_aix.cpp:458]: OK, I will use 64*K

File: src/hotspot/os/aix/os_aix.cpp (hunk 2 above)
Comment (JoKern65) [reply src/hotspot/os/aix/os_aix.cpp:459]: And what about the other guarantees in the same function after shmget() or shmat()?

Comment (tstuefe) [reply src/hotspot/os/aix/os_aix.cpp:459]: Yes, these are my faults. When I did them originally, I did not know better ( ~20 years ago). In new code, lets prefer assert.

File: test/hotspot/gtest/runtime/test_os.cpp (hunk 3 above)
Comment (JoKern65) [reply test/hotspot/gtest/runtime/test_os.cpp:47]: I hope I did it like you meant. Please verify.

File: test/hotspot/gtest/runtime/test_os_aix.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+#include "runtime/frame.inline.hpp"
+#include "runtime/globals.hpp"
+#include "runtime/os.inline.hpp"
+#include "runtime/thread.hpp"
+#include "runtime/threads.hpp"
+#include "utilities/align.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/macros.hpp"
+#include "utilities/ostream.hpp"
+#include "unittest.hpp"
```
Comment (tstuefe) [inline test/hotspot/gtest/runtime/test_os_aix.cpp:40]:
```suggestion
#include "runtime/os.inline.hpp"
#include "utilities/debug.hpp"
#include "utilities/globalDefinitions.hpp"
#include "unittest.hpp"
```

I would guess that is all includes you really need here

File: src/hotspot/os/aix/os_aix.cpp [hunk 5]
```diff
@@ -419,26 +428,30 @@ static void query_multipage_support() {
...
+            const size_t real_pagesize = os::Aix::query_pagesize(p);
+            if (real_pagesize != pagesize) {
+              log_warning(pagesize)("real page size (" SIZE_FORMAT_X ") differs.", real_pagesize);
+            } else {
+              can_use = true;
+            }
+            ::shmdt(p);
+          }
         }
-        ::shmdt(p);
```
Comment (tstuefe) [inline src/hotspot/os/aix/os_aix.cpp:441]: Are these just unrelated cleanups? I am not sure what this has to do with large-page mmap.

File: src/hotspot/os/aix/os_aix.cpp [hunk 6]
```diff
@@ -448,6 +461,16 @@ static void query_multipage_support() {
...
 
+    // Can we use mmap with 64K pages? (Should be available with AIX7.3 TL1)
+    {
+      void* p = mmap(NULL, 64*K, PROT_READ | PROT_WRITE, MAP_ANON_64K | MAP_ANONYMOUS | MAP_SHARED, -1, 0);
+      assert(p != (void*) -1, "mmap failed");
+      if (p != (void*) -1) {
+        g_multipage_support.can_use_64K_mmap_pages = (64*K == os::Aix::query_pagesize(p));
+        munmap(p, 64*K);
+      }
+    }
```
Comment (tstuefe) [inline src/hotspot/os/aix/os_aix.cpp:472]: Why MAP_SHARED?

File: src/hotspot/os/aix/os_aix.cpp [hunk 7]
```diff
@@ -2181,18 +2216,18 @@ void os::init(void) {
   //    and should be allocated with 64k pages.
   //
   // So, we do the following:
-  // LDR_CNTRL    can_use_64K_pages_dynamically       what we do                      remarks
-  // 4K           no                                  4K                              old systems (aix 5.2) or new systems with AME activated
-  // 4k           yes                                 64k (treat 4k stacks as 64k)    different loader than java and standard settings
+  // LDR_CNTRL    can_use_64K_pages_dynamically(mmap or shm)       what we do                      remarks
+  // 4K           no                                               4K                              old systems (aix 5.2) or new systems with AME activated
+  // 4k           yes                                              64k (treat 4k stacks as 64k)    different loader than java and standardsettings
```
Comment (tstuefe) [inline src/hotspot/os/aix/os_aix.cpp:2221]: standard settings (blank)

**Review by tstuefe — COMMENTED** [review]
Mostly good

File: test/hotspot/gtest/runtime/test_os.cpp (hunk 3 above)
Comment (JoKern65) [reply test/hotspot/gtest/runtime/test_os.cpp]: Yes you're right, I had tested this out beforehand, but was reluctant and not sure if I change something not obvious with this.

File: test/hotspot/gtest/runtime/test_os_aix.cpp (hunk 4 above)
Comment (JoKern65) [reply test/hotspot/gtest/runtime/test_os_aix.cpp:40]: Yes you're right, I had tested this out beforehand, but was reluctant and not sure if I change something not obvious with this.

File: src/hotspot/os/aix/os_aix.cpp (hunk 7 above)
Comment (tstuefe) [reply src/hotspot/os/aix/os_aix.cpp:2221]: In case this was not clear: I meant the typo. So, "s/standardsettings/standard settings" :)

File: src/hotspot/os/aix/os_aix.cpp [hunk 8]
```diff
@@ -96,6 +96,10 @@
 #include <sys/ioctl.h>
 #include <sys/ipc.h>
 #include <sys/mman.h>
+// sys/mman.h defines MAP_ANON_64K beginning with AIX7.3 TL1
+#ifndef MAP_ANON_64K
+#define MAP_ANON_64K  0x400
+#endif
```
Comment (tstuefe) [inline src/hotspot/os/aix/os_aix.cpp:102]:
```suggestion
// sys/mman.h defines MAP_ANON_64K beginning with AIX7.3 TL1
#ifndef MAP_ANON_64K
#define MAP_ANON_64K  0x400
#else
STATIC_ASSERT(MAP_ANON_64K == 0x400);
#endif
```
That adds a test, for when you *do* have the MAP_ANON_64K constant, that your assumption about its value (0x400) is correct. Needs utilities/debug.hpp, but you should include that anyway.

File: src/hotspot/os/aix/os_aix.cpp [hunk 9]
```diff
@@ -224,14 +228,15 @@ static struct {
   size_t textpsize;           // default text page size (LDR_CNTRL STACKPSIZE)
   bool can_use_64K_pages;     // True if we can alloc 64K pages dynamically with Sys V shm.
   bool can_use_16M_pages;     // True if we can alloc 16M pages dynamically with Sys V shm.
+  bool can_use_64K_mmap_pages;// True if we can alloc 64K pages dynamically with mmap.
```
Comment (tstuefe) [inline src/hotspot/os/aix/os_aix.cpp:231]: nit: blank after semicolon

File: src/hotspot/os/aix/os_aix.cpp [hunk 10]
```diff
@@ -419,26 +428,30 @@ static void query_multipage_support() {
...
-      } else {
-        // Attach and double check pageisze.
-        void* p = ::shmat(shmid, nullptr, 0);
-        ::shmctl(shmid, IPC_RMID, nullptr); // As early as possible!
-        guarantee0(p != (void*) -1); // Should always work.
-        const size_t real_pagesize = os::Aix::query_pagesize(p);
-        if (real_pagesize != pagesize) {
-          log_warning(pagesize)("real page size (" SIZE_FORMAT_X ") differs.", real_pagesize);
+      assert(shmid != -1, "shmget failed");
+      if (shmid != -1) {
```
Comment (tstuefe) [inline src/hotspot/os/aix/os_aix.cpp:432]: Ah, I see what you do here, you now want to handle shmid == -1 for the release build, because I wanted you to change guarantee->assert? Okay. That is very thorough.

File: src/hotspot/os/aix/os_aix.cpp (hunk 5 above)
Comment (tstuefe) [reply src/hotspot/os/aix/os_aix.cpp:441]: Forget my question, I now understand the (small) difference

File: src/hotspot/os/aix/os_aix.cpp (hunk 7 above)
Comment (JoKern65) [reply src/hotspot/os/aix/os_aix.cpp:2221]: Upps.

**Review by tstuefe — APPROVED** [review]
Okay

File: src/hotspot/os/aix/os_aix.cpp (hunk 6 above)
Comment (JoKern65) [reply src/hotspot/os/aix/os_aix.cpp:475]: Sorry Thomas, I oversaw your question.
I followed the the code of reserve_mmaped_memory(), where MAP_SHARED is used with the following comment:

```
  // Note: MAP_SHARED (instead of MAP_PRIVATE) needed to be able to
  // later use msync(MS_INVALIDATE) (see os::uncommit_memory).
  int flags = MAP_ANONYMOUS | MAP_SHARED;

```
I would like to have set the same flags for this test as used later on in real life.

Comment (tstuefe) [reply src/hotspot/os/aix/os_aix.cpp:475]: Okay.

File: src/hotspot/os/aix/os_aix.cpp [hunk 11]
```diff
@@ -67,6 +67,7 @@
 #include "services/attachListener.hpp"
 #include "services/runtimeService.hpp"
 #include "signals_posix.hpp"
+#include "utilities/debug.hpp"
```
Comment (TheRealMDoerr) [inline src/hotspot/os/aix/os_aix.cpp:70]: Should be sorted alphabetically.

File: src/hotspot/os/aix/os_aix.cpp [hunk 12]
```diff
@@ -96,6 +97,12 @@
 #include <sys/ioctl.h>
 #include <sys/ipc.h>
 #include <sys/mman.h>
+// sys/mman.h defines MAP_ANON_64K beginning with AIX7.3 TL1
+#ifndef MAP_ANON_64K
+  #define MAP_ANON_64K  0x400
```
Comment (TheRealMDoerr) [inline src/hotspot/os/aix/os_aix.cpp:102]: Extra whitespace.

File: test/hotspot/gtest/runtime/test_os_aix.cpp [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+#ifdef AIX
+
+#include "runtime/os.inline.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "unittest.hpp"
+#include <sys/mman.h>
+// sys/mman.h defines MAP_ANON_64K beginning with AIX7.3 TL1
+#ifndef MAP_ANON_64K
+  #define MAP_ANON_64K  0x400
```
Comment (TheRealMDoerr) [inline test/hotspot/gtest/runtime/test_os_aix.cpp:35]: Extra whitespace.

**Review by TheRealMDoerr — COMMENTED** [review]
LGTM. Only a few minor nits. Are you planning to change https://github.com/openjdk/jdk/blob/747e1e47f576b0ca3ac97d1deea87418e67ff2d1/src/hotspot/share/runtime/os.cpp#L1924 later?

File: src/hotspot/share/runtime/os.cpp [hunk 14]
```diff
@@ -1924,7 +1924,7 @@ char* os::attempt_reserve_memory_between(char* min, char* max, size_t bytes, siz
   // This is not reflected by os_allocation_granularity().
   // The logic here is dual to the one in pd_reserve_memory in os_aix.cpp
   const size_t system_allocation_granularity =
-    AIX_ONLY(os::vm_page_size() == 4*K ? 4*K : 256*M)
+    AIX_ONLY(os::vm_page_size() == 4*K ? 4*K : os::Aix::supports_64K_mmap_pages() ? 64*K : 256*M)
```
Comment (TheRealMDoerr) [inline src/hotspot/share/runtime/os.cpp:1927]: Please add braces or rewrite it. Does
`AIX_ONLY( (!os::Aix::supports_64K_mmap_pages() && os::vm_page_size() == 64*K) ? 256*M : ) os::vm_allocation_granularity()` work?

**Review by TheRealMDoerr — APPROVED** [review]
Thanks! LGTM. Maybe @tstuefe can take a look at the latest changes, too.

**Review by tstuefe — APPROVED** [review]
Very good.

**Review by MBaesken — APPROVED** [review]
